packages/astro/src/core/build/buildPipeline.ts

```diff
@@ -0,0 +1,211 @@
+import { Pipeline } from '../pipeline.js';
+import type { BuildInternals } from './internal';
+import type { PageBuildData, StaticBuildOptions } from './types';
+import { ASTRO_PAGE_RESOLVED_MODULE_ID } from './plugins/plugin-pages.js';
+import { RESOLVED_SPLIT_MODULE_ID } from './plugins/plugin-ssr.js';
+import { ASTRO_PAGE_EXTENSION_POST_PATTERN } from './plugins/util.js';
+import type { SSRManifest } from '../app/types';
+import type { AstroConfig, AstroSettings, RouteType, SSRLoadedRenderer } from '../../@types/astro';
+import { getOutputDirectory, isServerLikeOutput } from '../../prerender/utils.js';
+import type { EndpointCallResult } from '../endpoint';
+import { createEnvironment } from '../render/index.js';
+import { BEFORE_HYDRATION_SCRIPT_ID } from '../../vite-plugin-scripts/index.js';
+import { createAssetLink } from '../render/ssr-element.js';
+import type { BufferEncoding } from 'vfile';
+
+/**
+ * This pipeline is responsible to gather the files emitted by the SSR build and generate the pages by executing these files.
+ */
+export class BuildPipeline extends Pipeline {
+	#internals: BuildInternals;
+	#staticBuildOptions: StaticBuildOptions;
+	#manifest: SSRManifest;
+	#currentEndpointBody?: {
+		body: string | Uint8Array;
+		encoding: BufferEncoding;
+	};
+
+	constructor(
+		staticBuildOptions: StaticBuildOptions,
+		internals: BuildInternals,
+		manifest: SSRManifest
+	) {
+		const ssr = isServerLikeOutput(staticBuildOptions.settings.config);
+		super(
+			createEnvironment({
+				adapterName: manifest.adapterName,
+				logging: staticBuildOptions.logging,
+				mode: staticBuildOptions.mode,
+				renderers: manifest.renderers,
+				clientDirectives: manifest.clientDirectives,
+				compressHTML: manifest.compressHTML,
+				async resolve(specifier: string) {
+					const hashedFilePath = manifest.entryModules[specifier];
+					if (typeof hashedFilePath !== 'string') {
+						// If no "astro:scripts/before-hydration.js" script exists in the build,
+						// then we can assume that no before-hydration scripts are needed.
+						if (specifier === BEFORE_HYDRATION_SCRIPT_ID) {
```

Comment:
I think this check is the reason that so many hybrid render builds are failing, 
in `packages/src/core/build/plugins/plugin-manifest.ts:238` you are setting `entryModules[BEFORE_HYDRATION_SCRIPT_ID] = ''`, then here you are checking `if (typeof hashedFilePath !== 'string')` and inside that check you are checking `if (specifier === BEFORE_HYDRATION_SCRIPT_ID)`, but as you have already set the `BEFORE_HYDRATION_SCRIPT_ID` in `plugin-manifest.ts` so this check is never called.

as a result, while resolving `BEFORE_HYDRATION_SCRIPT_ID` in `packages/astro/src/runtime/server/hydration.ts:157` we get `"/"` instead of `""`, for which this check passes and `before-hydration-url` is generated.
```js
if (beforeHydrationUrl.length) {
		island.props['before-hydration-url'] = beforeHydrationUrl;
	}
```

So the potential fixes could be to change the function to be
```js
async resolve(specifier: string) {
					const hashedFilePath = manifest.entryModules[specifier];
					// If no "astro:scripts/before-hydration.js" script exists in the build,
					// then we can assume that no before-hydration scripts are needed.
					if (specifier === BEFORE_HYDRATION_SCRIPT_ID) {
						return '';
					}
					if (typeof hashedFilePath !== 'string') {
						throw new Error(`Cannot find the built path for ${specifier}`);
					}
					return createAssetLink(hashedFilePath, manifest.base, manifest.assetsPrefix);
				},
```
or change the check at `packages/astro/src/runtime/server/hydration.ts:157` to be
```js
if (beforeHydrationUrl.length !== 0 && beforeHydrationUrl !== "/") {
		island.props['before-hydration-url'] = beforeHydrationUrl;
	}
```

I am down to make a PR to fix the issue (#8379 and many more), just waiting for confrimations from the maintainers